### PR TITLE
Fleet add repository form update - #5197

### LINKED ISSUE #5197: Fleet add repository form update.
**Detailed Description**
Step 1
image

Step 2
image

**Context**
TBC: Part of wider changed to fleet dashboard area.

### PR DESCRIPTION
**Pull request by Shavindra**
### Summary
Fixes #5197 - Fleet add repository form update.

### Changes
Split the form so that it has 2 steps instead of one

### Technical notes summary
- Added the step functionalities from `.components/Wizzard.vue` to the `/components/CruResource.vue`

#### Notes:
- All the existing form functionality including error handling should work as it was doing before
- Only difference is Git repo info and Target info is split into 2 steps
- Should still be able to edit Yaml and filled in information should appear

#### Steps
- Navigate to Continuous Delivery > Git Repos 
- Click on `Add repository`
- Use your own git repo or  examples from  https://github.com/rancher/fleet-examples to add a repo.

### Areas which could experience regressions
- Form functionality and fields
- Resource creation e.g. Cluster creation, cluster workspace creation

### Screenshot/Video

https://user-images.githubusercontent.com/1387263/163439071-ea1b84ac-bd39-4b44-a717-166f4cc7ca67.mov

### CODE REVIEW

**Review by richard-cox — CHANGES_REQUESTED** [review]
`CruResourceWizard` component contains large blocks of copy & paste from both `CruResource` and `Wizard` 

If we go with the idea of a `CruResourceWizard` we need to ensure it's a lightweight wrapper around the existing Wizard without the duplicated code of either. Button integration can be added to either inject buttons into the wizard or `CruResource`.

**Comment by Shavindra** [comment]
@richard-cox yes, I didn't want to refactor existing code. I wasn't sure if its going to break anything. I was going to wait after the merge to go in and refactor to inject buttons. Let me know if you still like to do this as part of this PR? or after?

**Comment by richard-cox** [comment]
@Shavindra There's a balance between potentially dangerous changes and technical debt, I think in this case the scales are in favour of the changes.

If the practise of `slots` are stuck to then it should minimise the danger, so the CruResourceWizard would override `slots` in the Wizard for the header and buttons. The slot for the buttons could provide a lot of the plumbing (not sure of names but things like canNext, Next(), canPrevious(), etc)

**Comment by richard-cox** [comment]
Update on status, briefly went through this with Shavin on Friday and there's some tweaks still to do

**Review by richard-cox — CHANGES_REQUESTED** [review]
There's still a lot of copy and paste from the Wizard throughout here. 

Anything containing internal wizard logic like activeStepIndex, showPrevious, stepsLoaded, goToStep etc should be accessed via the wizard itself. 

Any properties left that are absolutely vital should be defined in a mixin to keep them nicely separate (this isn't a pattern that's been well applied, but is done well in the `SortableTable`).

File: components/CruResource.vue [hunk 1]
```diff
@@ -89,13 +96,65 @@ export default {
...
       default: null,
-    }
+    },
+    steps: {
+      type:     Array,
+      default: () => []
+    },
+
+    // Initial step to show when Wizard loads.
+    initStepIndex: {
```
Comment (richard-cox) [inline components/CruResource.vue:106]: not used, can be removed

File: components/CruResource.vue [hunk 2]
```diff
@@ -89,13 +96,65 @@ export default {
...
+    },
+
+    // if true, allow users to navigate back to the first step of the Wizard
+    // if false, only way back to step 1 is to cancel and undo all configuration
+    editFirstStep: {
+      type:    Boolean,
+      default: false
+    },
+
+    showBanner: {
```
Comment (richard-cox) [inline components/CruResource.vue:118]: not used, can be removed

File: components/CruResource.vue [hunk 3]
```diff
@@ -89,13 +96,65 @@ export default {
...
+      type:    Boolean,
+      default: true,
+    },
+
+    // whether or not to show the overall title/image on left of banner header in first step
+    initialTitle: {
```
Comment (richard-cox) [inline components/CruResource.vue:124]: not used, can be removed

File: components/CruResource.vue [hunk 4]
```diff
@@ -257,6 +316,10 @@ export default {
       this.$refs.save.clicked();
     },
 
+    handleStepChange(e) {
```
Comment (richard-cox) [inline components/CruResource.vue:319]: given the changes to the wizard slot/templates  i don't think this is needed anymore (along with the local copy of `activeStep` - when used in a template that should come from a slot)

File: components/Wizard.vue [hunk 5]
```diff
@@ -184,6 +189,7 @@ export default {
       this.activeStep = selected;
 
       this.$emit('next', { step: selected });
+      this.$emit('stepChange', { step: selected });
```
Comment (richard-cox) [inline components/Wizard.vue:192]: given comment above hopefully this can be removed

**Review by richard-cox — CHANGES_REQUESTED** [review]

**Review by richard-cox — CHANGES_REQUESTED** [review]
App Charts install wizard
  - there's no space now between header and content
  image


Git Repo Wizard
- You'll need to rebase with master. There was a bug which resulted in the cru resource error section to always show (including it's margin). Manually removing the section in this branch (as if it were master) makes the header look odd
  image
  image
- For this usage of the wizard header, where there are no icons, it consumes a lot of space. The `Wizard` should provide a way to override the .header > .title min-height . The cru resource should set this to something around 60px (the steps on the right will still need to look ok)
- The first step is titled `Finish: Step 1`. The verb should be linked to the change of resource like in the header above. Though given it's there I don't think it needs repeating here so can be removed
  image
- There's a lot of content in step 1, user would need to scroll in most windows which isn't very step-ish. So would suggest creating an additional step. Maybe move git and helm auth fields into a `Repo Auth` step? To confirm with Neil.
- There's 2 fields marked as required on step one and the `Next` button is enabled when either are empty. To fix this you can add `ready: false` to the `step` 1 object.

**Review by richard-cox — CHANGES_REQUESTED** [review]
App Charts install wizard
- "there's no space now between header and content". Header & Content being the vertical contexts. So...
  image
  vs 
  image
- The footer is now broken
  image

Git Repo Wizard
- The vertical alignment for the step details part (to high) and the step breadcrumbs (too low) are now out 
  image
- "The first step is titled Finish: Step 1. The verb should be linked to the change of resource like in the header above. Though given it's there I don't think it needs repeating here so can be removed"
  - The actual line is `<h2>{{ t(`asyncButton.${headerMode || finishMode}.action`) }}: {{ t('wizard.step', {number:activeStepIndex+1}) }}</h2>` so passing in a `headerMode` matching the create/edit context should work

**Review by richard-cox — CHANGES_REQUESTED** [review]
App Charts install wizard
- "there's no space now between header and content". Header & Content being the vertical contexts. So...
  image
  vs 
  image
- The footer is now broken
  image

Git Repo Wizard
- The vertical alignment for the step details part (to high) and the step breadcrumbs (too low) are now out 
  image
- "The first step is titled Finish: Step 1. The verb should be linked to the change of resource like in the header above. Though given it's there I don't think it needs repeating here so can be removed"
  - The actual line is 
  - ```<h2>{{ t(`asyncButton.${headerMode || finishMode}.action`) }}: {{ t('wizard.step', {number:activeStepIndex+1}) }}</h2>``` so passing in a `headerMode` matching the create/edit context should work
- I entered a git and helm auth on step 1, hit next for step2, hit previous back to step1 and the controls for both had been reset. looks like the same issue when editing an existing repo with both settings
- When editing an existing repo the next button on step one didn't become enabled until there was a change of any value

**Review by richard-cox — CHANGES_REQUESTED** [review]
Git Repo Wizard
- The alignment in the header is still out
  image
  - The vertical alignment for the step details part is now too low, it needs to be centered
  - the step breadcrumbs are unchanged and too low. They should be centered as well with the line leading between the dots.
- The first step is still titled "Finish: Step 1". See comment in https://github.com/rancher/dashboard/pull/5625#pullrequestreview-990102151
- git and helm auth on step 1, is still lost on return. See comment in https://github.com/rancher/dashboard/pull/5625#pullrequestreview-990102151

**Review by richard-cox — CHANGES_REQUESTED** [review]
- Loading the create repo page throws a full page dev error due to `AUTH_TYPE` missing in data in `SelectOrCreateAuthSecret`
- There's an issue when viewing the github repo `Config` tab (see below). That tab shows the resource's `edit` component but in `VIEW` mode (all components disabled, no cancel/save buttons). As a low-bar solution the wizard within CruResouce should also take this in to account (no content left of the breadcrumbs, no edit as YAML, form controls disabled by passing on `mode`), 
   image

File: shell/edit/fleet.cattle.io.gitrepo.vue [hunk 6]
```diff
@@ -117,6 +138,16 @@ export default {
       return _SPECIFY;
     },
 
+    addRepositorySteps() {
```
[the remark this answers is not in the document]
Comment (Shavindra) [reply shell/edit/fleet.cattle.io.gitrepo.vue:141]: The steps are dynamic

**Review by richard-cox — CHANGES_REQUESTED** [review]

File: shell/components/CruResourceFooter.vue [hunk 7]
```diff
@@ -74,7 +74,7 @@ export default {
     <slot name="footer-prefix" />
     <slot name="cancel">
       <button
-        v-if="!isView && showCancel"
+        v-if="isView || showCancel"
```
[the remark this answers is not in the document]
Comment (richard-cox) [reply shell/components/CruResourceFooter.vue:77]: That's fine. This is on the config tab and the other edit pages in this context don't have the cancel buttons either.

Going forward this should hopefully change anyway. The wizard should be used for create flows and the traditional tab way for edit flows (the edit flow would then replace this wizard in the config tab). For this round though we're falling back on the wizard.

**Review by richard-cox — APPROVED** [review]
LGTM! Remember to squash merge

**Pull request merged**